libs/@local/hash-validation/src/property_type.rs

```diff
@@ -0,0 +1,702 @@
+                PropertyValidationError::InvalidPropertyKey { key: key.clone() }
+            })?;
+            if let Some(object_schema) = self.properties().get(&key) {
+                if let Err(report) = object_schema.validate_value(property, provider).await {
+                    extend_report!(
+                        status,
+                        report.change_context(PropertyValidationError::InvalidProperty {
+                            key: key.clone(),
+                        })
+                    );
+                }
+            } else {
+                extend_report!(
+                    status,
+                    PropertyValidationError::UnexpectedProperty { key: key.clone() }
+                );
+            }
+        }
+
+        for required_property in self.required() {
+            if !value.contains_key(required_property.as_str()) {
+                extend_report!(
+                    status,
+                    PropertyValidationError::MissingRequiredProperty {
+                        key: required_property.clone(),
+                    }
+                );
+            }
+        }
+
+        status
+    }
+}
+
+impl<P, const MIN: usize> Schema<HashMap<BaseUrl, JsonValue>, P>
+    for Object<ValueOrArray<PropertyTypeReference>, MIN>
+where
+    P: OntologyTypeProvider<PropertyType> + OntologyTypeProvider<DataType> + Sync,
+{
+    type Error = PropertyValidationError;
+
+    async fn validate_value<'a>(
+        &'a self,
+        value: &'a HashMap<BaseUrl, JsonValue>,
+        provider: &'a P,
+    ) -> Result<(), Report<Self::Error>> {
+        let mut status: Result<(), Report<Self::Error>> = Ok(());
+
+        for (key, property) in value {
+            if let Some(object_schema) = self.properties().get(key) {
+                if let Err(report) = object_schema.validate_value(property, provider).await {
+                    extend_report!(status, report);
+                }
+            } else {
+                extend_report!(
+                    status,
+                    PropertyValidationError::UnexpectedProperty { key: key.clone() }
+                );
+            }
+        }
+
+        for required_property in self.required() {
+            if !value.contains_key(required_property) {
+                extend_report!(
+                    status,
+                    PropertyValidationError::MissingRequiredProperty {
+                        key: required_property.clone(),
+                    }
+                );
+            }
+        }
+
+        status
+    }
+}
+
+impl<P> Schema<JsonValue, P> for PropertyValues
+where
+    P: OntologyTypeProvider<PropertyType> + OntologyTypeProvider<DataType> + Sync,
+{
+    type Error = PropertyValidationError;
+
+    fn validate_value<'a>(
+        &'a self,
+        value: &'a JsonValue,
+        provider: &'a P,
+    ) -> Pin<Box<dyn Future<Output = Result<(), Report<Self::Error>>> + Send + '_>> {
```

Comment:
the beauty of recursive async functions

Reply:
Haha, yes. I just spoke with @nonparibus today about recursive `async` functions in Rust and why this is so special. 😄